packages/styled-components/src/sheet/HoistedTag.js

```diff
@@ -0,0 +1,50 @@
+// @flow
+
+import type { HoistedTag, GroupedTag, Tag } from './types';
+import { DefaultGroupedTag } from './GroupedTag';
+import { WindowedTag } from './WindowedTag';
+
+const hoistedRuleRe = /\s*@(font-face|import)/;
+
+export class DefaultHoistedTag implements HoistedTag {
+  hoistedTag: GroupedTag;
+
+  normalTag: GroupedTag;
+
+  length: number;
+
+  constructor(tag: Tag) {
+    this.hoistedTag = new DefaultGroupedTag(new WindowedTag(tag));
+    this.normalTag = new DefaultGroupedTag(new WindowedTag(tag));
+    this.length = 0;
+  }
+
+  insertRules(group: number, rules: string[]): void {
+    const hoistedRules = [];
+    const normalRules = [];
+
+    for (let i = 0, l = rules.length; i < l; i++) {
+      const rule = rules[i];
+      if (hoistedRuleRe.test(rule)) {
+        hoistedRules.push(rule);
+      } else {
+        normalRules.push(rule);
+      }
+    }
+
+    this.normalTag.insertRules(group, normalRules);
+    this.hoistedTag.insertRules(group, hoistedRules);
+    this.length = this.normalTag.length;
+  }
+
+  clearGroup(group: number) {
+    this.normalTag.clearGroup(group);
+    this.hoistedTag.clearGroup(group);
+    this.length = this.normalTag.length;
+  }
+
+  // TODO: Needs to be replaced to keep hoisted ordering
```

Comment:
@kitten  - Can you expound a little on what this means?

Reply:
This command isn’t yet in use. Instead the Rehydration sheet actually just calls into the hoisted and normal tags directly which is bad.

Basically, the output logic should be as close to every tag as possible. To the point where calling toString on the sheet’s HoistedTag directly gives you the right CSS output.

This will require a couple of changes however:

- the registered names need to either be removed from the SSR output or moved to the tags. It may be possible to generate them on the fly, but that may require even more changes and comes at the cost of sacrificing perf that we fought for 😅
- all hoisted rules must come before normal rules (that’s already done, but hacky as mentioned before)
- the global rules either need special “marker rules” or the marker rules need to be changed so that they don’t interfere with global rules. We should still however be using marker rules if possible. Parsing and splitting the CSS manually, which is what we did before, was pretty slow

Reply:
@kitten - I gave this a shot in my PR #2775 
I went with making sure the hoisted rules are output first by diverging the HoistedTag from the GroupedTag and then I ended up ignoring the hoisted rules on Rehydration (possibly a bad thing, but I didn't see any negatives in the sandbox, tests, or my app I was testing in locally). I'd appreciate a review of that PR when you get a chance, since I think I hit everything except the "code could probably be cleaned up" notion.